### PR TITLE
Move goto_convert files

### PR DESCRIPTION
**Pull request by kroening**
The `goto_convertt` class and associated helpers convert a C parse tree into a set of GOTO functions.  They are specific to C, and hence, should be in the `ansi-c/` directory.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/assignments_from_json.cpp [hunk 1]
```diff
@@ -16,7 +16,7 @@ Author: Diffblue Ltd.
 #include <util/symbol_table_base.h>
 #include <util/unicode.h>
 
-#include <goto-programs/allocate_objects.h>
+#include <ansi-c/allocate_objects.h>
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/assignments_from_json.cpp]: Is this one really C specific? It appears to be used by JBMC, and perhaps more so than anywhere else?

**Comment by tautschnig** [comment]
I have a couple of concerns:
* The ansi-c folder is starting to become unmanageable. At bare minimum I'd suggest that we have a new subfolder `ansi-c/goto-conversion/` (or some similar name) to hold those files.
* As indicated by my comment, but perhaps applicable in more cases: are all those files really all about goto conversion?
* Is it really correct to assert that only C parse trees can be converted? Isn't it rather the case that we actually have a further intermediate language of `codet` statements?
* Do we perhaps need to become a lot more clinical on where we use `goto_convert`? As this PR documents, we make use of it in a number of places well outside the C front-end. To that effect, I'd like to put up for debate whether we do such cleanup before moving files.

File: jbmc/src/java_bytecode/assignments_from_json.cpp (hunk 1 above)
Comment (kroening) [reply jbmc/src/java_bytecode/assignments_from_json.cpp]: I'd say it tries to do too much. Object construction is clearly language specific, and this should get split up.

**Comment by kroening** [comment]
> * The ansi-c folder is starting to become unmanageable. At bare minimum I'd suggest that we have a new subfolder `ansi-c/goto-conversion/` (or some similar name) to hold those files.

Ok, will do.

> * As indicated by my comment, but perhaps applicable in more cases: are all those files really all about goto conversion?

No, there's also object construction, but that appears to be intertwined with the conversion. As said in the comment about it, it's also language specific, and hence should move out of the goto-programs folder.

> * Is it really correct to assert that only C parse trees can be converted? Isn't it rather the case that we actually have a further intermediate language of `codet` statements?

This is getting reduced, and I'd like to get rid of it altogether.

> * Do we perhaps need to become a lot more clinical on where we use `goto_convert`? As this PR documents, we make use of it in a number of places well outside the C front-end. To that effect, I'd like to put up for debate whether we do such cleanup before moving files.

We realistically won't do the cleanup. For symtab2gb, this lives outside of the codebase. For jbmc, we don't have the person to do it.

**Comment by tautschnig** [comment]
> > * The ansi-c folder is starting to become unmanageable. At bare minimum I'd suggest that we have a new subfolder `ansi-c/goto-conversion/` (or some similar name) to hold those files.
> 
> Ok, will do.
[...]

Happy to approve once this is done: I think this goes a long way towards making all of my other concerns easy to track (and, if resources ever permit: fix).

**Comment by kroening** [comment]
Now with ansi-c/goto-conversion subdirectory.

**Review by tautschnig — APPROVED** [review]

File: src/linking/module_dependencies.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
+ansi-c
```
Comment (peterschrammel) [inline src/linking/module_dependencies.txt:1]: ❓ Isn't linking a pure GOTO thing? Why is there a dependency on ansi-c now? This seems to indicate that either some functionality has been moved that should not be moved or that some functionality is used in other modules that shouldn't be used there.

Comment (kroening) [reply src/linking/module_dependencies.txt:1]: The culprit for this is `recreate_initialize_function`. I'd like to kill it, and one benefit would be to remove this coupling.

Comment (tautschnig) [reply src/linking/module_dependencies.txt:1]: Being the one who introduced this function in the first place: can you share some ideas on how to replace it?

Comment (kroening) [reply src/linking/module_dependencies.txt:1]: As mentioned out-of-band, simply create the goto instructions directly. This should be done separately.

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**